quixstreams/rowconsumer/buffering.py

```diff
@@ -0,0 +1,403 @@
+        - The Consumer calls `.pop()` to get the next message to be processed.
+            If multiple partitions with the same number are assigned, the message will be popped from the partition with the smallest next timestamp, providing in-order reads.
+        - The Consumer calls `.pause_full()` and `.resume_empty()` methods to balance the reads across all partitions.
+
+        :param max_partition_buffer_size: the maximum size of the individual topic partition buffer when the buffer is considered full.
+            It is a soft limit, and it may be exceeded in some cases.
+            When individual buffer exceeds this limit, its TP can be paused to let other partitions to be consumed too.
+        """
+        self._partition_groups: dict[int, PartitionBufferGroup] = {}
+        self._max_partition_buffer_size = max_partition_buffer_size
+
+    def on_assign(self, topic_partitions: list[TopicPartition]):
+        """
+        Assign new partitions to the buffer.
+
+        :param topic_partitions: list of `confluent_kafka.TopicPartition`.
+        """
+        # Sort partitions by their number to process them
+        # in a fixed order when popping the items over
+        topic_partitions = sorted(topic_partitions, key=lambda t: t.partition)
+        for tp in topic_partitions:
+            partition_group = self._partition_groups.setdefault(
+                tp.partition,
+                PartitionBufferGroup(
+                    partition=tp.partition, max_size=self._max_partition_buffer_size
+                ),
+            )
+            partition_group.assign_partition(topic=tp.topic)
+
+    def on_revoke(self, topic_partitions: list[TopicPartition]):
+        """
+        Drop the partitions from the buffer.
+
+        :param topic_partitions: list of `confluent_kafka.TopicPartition`.
+        """
+        for tp in topic_partitions:
+            partition_group = self._partition_groups.get(tp.partition)
+            if partition_group is not None:
+                partition_group.revoke_partition(topic=tp.topic)
+
+    def feed(
+        self,
+        messages: Iterable[SuccessfulConfluentKafkaMessageProto],
+        high_watermarks: dict[tuple[str, int], int],
+    ):
+        """
+        Feed new batch of messages to the buffer.
+
+        :param messages: an iterable with successful `confluent_kafka.Message` objects (`.error()` is expected to be None).
+        :param high_watermarks: a dictionary with high watermarks for all assigned topic partitions.
+        """
+        for message in messages:
+            partition_group = self._partition_groups[message.partition()]
+            partition_group.append(message=message)
+
+        for partition_group in self._partition_groups.values():
+            group_watermarks = {
+                topic: watermark
+                for (topic, partition), watermark in high_watermarks.items()
+                if partition == partition_group.partition
```

Comment:
If you accept the suggestion in `_feed_buffer`, then this is an accompanying change
```suggestion
                tp.topic: watermark
                for tp, watermark in high_watermarks.items()
                if tp.partition == partition_group.partition
```